contrib/graphite/blueflood.py

```diff
@@ -43,37 +77,133 @@ def __init__(self, config=None):
 
     self.tenant = tenant
     self.bf_query_endpoint = urls[0]
+    self.enable_submetrics = enable_submetrics
+    self.submetric_aliases = submetric_aliases
+    self.client = BluefloodClient(self.bf_query_endpoint, self.tenant, 
+                                  self.enable_submetrics, self.submetric_aliases)
+    print("BF finder submetrics enabled: ", enable_submetrics)
+
+  def complete(self, metric, query_depth):
+    metric_parts = metric.split('.')
+    if query_depth < len(metric_parts):
+      return False
+    if self.enable_submetrics and (query_depth == len(metric_parts)):
+      return False
+    return True
+
+  def make_request(self, url, payload, headers):
+    if auth.is_active():
+      headers['X-Auth-Token'] = auth.get_token(False)
+    r = requests.get(url, params=payload, headers=headers)
+    if r.status_code == 401 and auth.is_active():
+      headers['X-Auth-Token'] = auth.get_token(True)
+      r = requests.get(url, params=payload, headers=headers)
+    return r
+
+  def find_metrics_endpoint(self, endpoint, tenant):
+    return "%s/v2.0/%s/metrics/search" % (endpoint, tenant)
+
+  def find_metrics(self, query):
+    print "BluefloodClient.find_metrics: " + str(query)
+    payload = {'query': query}
+    headers = auth.headers()
+    endpoint = self.find_metrics_endpoint(self.bf_query_endpoint, self.tenant)
+    r = self.make_request(endpoint, payload, headers)
+
+    if r.status_code == 200:
+        return [m['metric'] for m in r.json()]
+    else:
+      return []
+
+  def find_nodes_with_submetrics(self, query):
+    # By definition, when using submetrics, the names of all Leafnodes must end in a submetric alias
+    # BF doesn't know about the submetric aliases and so the aliases, (or globs corresponding to them,)
+    # must be truncated before querying BF.
+    # There are two cases above:
+    # 1. When you want a list of valid submetric aliases, i.e. a complete metric name followed by ".*"
+    # 2. When you have a query contains a valid submetric alias, i.e a complete metric name followed
+    #    by a valid submetric alias like "._avg"
+    # Every submetric leaf node is covered by one of the above two cases.  Everything else is a 
+    # branch node.
+
+    query_parts = query.pattern.split('.')
+    query_depth = len(query_parts)
+
+    if query_depth > 1:
+      #  In this if clause we are searching for complete metric names
+      #  followed by a ".*". If so, we are requesting a list of submetric
+      #  names, so return leaf nodes for each submetric alias that is
+      #  enabled
+      if query_parts[-1] == '*':
+        for metric in self.find_metrics('.'.join(query_parts[0:-1])):
```

Comment:
If the query end with a `*` but its not for a leaf node, it will make a call to `find_metrics` here and because its not complete, it will make a call again to `find_metrics` to return a branch node. Are these actual API calls or the data returned is cached?